### PR TITLE
dynamically deserialize request log event payload `request` field

### LINKED ISSUE #702: Stripe-cli 1.6.0 to 1.6.4 ignores valid payloads and sees them as malformed events
### Issue
Webhook payloads for customer.subscription.updated were filtered out by Stripe CLI version 1.6.4 instead of being forwarded to the desired endpoint

### Expected Behavior
web hooks get forwarded to configured endpoint

### Steps to reproduce
Set your default api version to 2015-09-03
Use stripe-cli any version after and including 1.6.0
Set Stripe CLI to listen to all all web hooks and forward them somewhere
Make sure logging level is set to debug
Update a subscription via the dashboard
Look for the incoming payload followed by time="Wed, 07 Jul 2021 14:20:12 UTC" level=debug msg="Received malformed event from Stripe, ignoring"

### Traceback
```
time="Wed, 07 Jul 2021 14:20:10 UTC" level=debug msg="Incoming message" message="{"webhook_id":"1625667607-wh_BjYS4iHtNOuaoS","webhook_conversation_id":"1625667607-wc_LvR5FnwsjOWiVG","event_payload":"{\n  \"id\": \"evt_1JAbbfCJZWINMVesDB7BL9Gd\",\n  \"object\": \"event\",\n  \"api_version\": \"2020-08-27\",\n  \"created\": 1625667607,\n  \"data\": {\n    \"object\": {\n      \"id\": \"sub_Jo9k3qUoCL2CdR\",\n      \"object\": \"subscription\",\n      \"application_fee_percent\": null,\n      \"automatic_tax\": {\n        \"enabled\": false\n      },\n      \"billing_cycle_anchor\": 1628734656,\n      \"billing_thresholds\": null,\n      \"cancel_at\": null,\n      \"cancel_at_period_end\": false,\n      \"canceled_at\": null,\n      \"collection_method\": \"charge_automatically\",\n      \"created\": 1625651544,\n      \"current_period_end\": 1628734656,\n      \"current_period_start\": 1625651544,\n      \"customer\": \"cus_Jo9KjcY44fGzj1\",\n      \"days_until_due\": null,\n      \"default_payment_method\": null,\n      \"default_source\": null,\n      \"default_tax_rates\": [\n\n      ],\n      \"discount\": null,\n      \"ended_at\": null,\n      \"items\": {\n        \"object\": \"list\",\n        \"data\": [\n          {\n            \"id\": \"si_Jo9kt4sglulMkz\",\n            \"object\": \"subscription_item\",\n            \"billing_thresholds\": null,\n            \"created\": 1625651545,\n            \"metadata\": {\n              \"orderCount\": \"1\",\n              \"lastOrderId\": \"1033825\"\n            },\n            \"plan\": {\n              \"id\": \"price_1HpECzCJZWINMVesSprmNjl5\",\n              \"object\": \"plan\",\n              \"active\": true,\n              \"aggregate_usage\": null,\n              \"amount\": 1680,\n              \"amount_decimal\": \"1680\",\n              \"billing_scheme\": \"per_unit\",\n              \"created\": 1612534011,\n              \"currency\": \"gbp\",\n              \"interval\": \"day\",\n              \"interval_count\": 28,\n              \"livemode\": false,\n              \"metadata\": {\n                \"boxes\": \"1\",\n                \"boxSize\": \"28\",\n                \"dailyConsumption\": \"1\",\n                \"baseSku\": \"TURME\",\n                \"retired\": \"0\",\n                \"productVariant\": \"TURME28\"\n              },\n              \"nickname\": \"2021 Turmeric+ 28\",\n              \"product\": \"prod_IQ4LQ4YFeqYB20\",\n              \"tiers_mode\": null,\n              \"transform_usage\": null,\n              \"trial_period_days\": 24,\n              \"usage_type\": \"licensed\"\n            },\n            \"price\": {\n              \"id\": \"price_1HpECzCJZWINMVesSprmNjl5\",\n              \"object\": \"price\",\n              \"active\": true,\n              \"billing_scheme\": \"per_unit\",\n              \"created\": 1612534011,\n              \"currency\": \"gbp\",\n              \"livemode\": false,\n              \"lookup_key\": null,\n              \"metadata\": {\n                \"boxes\": \"1\",\n                \"boxSize\": \"28\",\n                \"dailyConsumption\": \"1\",\n                \"baseSku\": \"TURME\",\n                \"retired\": \"0\",\n                \"productVariant\": \"TURME28\"\n              },\n              \"nickname\": \"2021 Turmeric+ 28\",\n              \"product\": \"prod_IQ4LQ4YFeqYB20\",\n              \"recurring\": {\n                \"aggregate_usage\": null,\n                \"interval\": \"day\",\n                \"interval_count\": 28,\n                \"trial_period_days\": 24,\n                \"usage_type\": \"licensed\"\n              },\n              \"tiers_mode\": null,\n              \"transform_quantity\": null,\n              \"type\": \"recurring\",\n              \"unit_amount\": 1680,\n              \"unit_amount_decimal\": \"1680\"\n            },\n            \"quantity\": 1,\n            \"subscription\": \"sub_Jo9k3qUoCL2CdR\",\n            \"tax_rates\": [\n\n            ]\n          }\n        ],\n        \"has_more\": false,\n        \"total_count\": 1,\n        \"url\": \"/v1/subscription_items?subscription=sub_Jo9k3qUoCL2CdR\"\n      },\n      \"latest_invoice\": \"in_1JAXQbCJZWINMVes1Sjxxxh6\",\n      \"livemode\": false,\n      \"metadata\": {\n        \"shipping\": \"true\",\n        \"shippingUpgraded\": \"false\",\n        \"source\": \"magento\",\n        \"orderCount\": \"1\",\n        \"lastOrderId\": \"1033825\"\n      },\n      \"next_pending_invoice_item_invoice\": null,\n      \"pause_collection\": null,\n      \"pending_invoice_item_interval\": null,\n      \"pending_setup_intent\": null,\n      \"pending_update\": null,\n      \"plan\": {\n        \"id\": \"price_1HpECzCJZWINMVesSprmNjl5\",\n        \"object\": \"plan\",\n        \"active\": true,\n        \"aggregate_usage\": null,\n        \"amount\": 1680,\n        \"amount_decimal\": \"1680\",\n        \"billing_scheme\": \"per_unit\",\n        \"created\": 1612534011,\n        \"currency\": \"gbp\",\n        \"interval\": \"day\",\n        \"interval_count\": 28,\n        \"livemode\": false,\n        \"metadata\": {\n          \"boxes\": \"1\",\n          \"boxSize\": \"28\",\n          \"dailyConsumption\": \"1\",\n          \"baseSku\": \"TURME\",\n          \"retired\": \"0\",\n          \"productVariant\": \"TURME28\"\n        },\n        \"nickname\": \"2021 Turmeric+ 28\",\n        \"product\": \"prod_IQ4LQ4YFeqYB20\",\n        \"tiers_mode\": null,\n        \"transform_usage\": null,\n        \"trial_period_days\": 24,\n        \"usage_type\": \"licensed\"\n      },\n      \"quantity\": 1,\n      \"schedule\": null,\n      \"start_date\": 1625651544,\n      \"status\": \"trialing\",\n      \"transfer_data\": null,\n      \"trial_end\": 1628734656,\n      \"trial_start\": 1625651544\n    },\n    \"previous_attributes\": {\n      \"billing_cycle_anchor\": 1627612809,\n      \"current_period_end\": 1627612809,\n      \"trial_end\": 1627612809\n    }\n  },\n  \"livemode\": false,\n  \"pending_webhooks\": 6,\n  \"request\": {\n    \"id\": \"req_oVXbEfceo8VdC1\",\n    \"idempotency_key\": null\n  },\n  \"type\": \"customer.subscription.updated\"\n}","http_headers":{"Content-Type":"application/json; charset=utf-8","Cache-Control":"no-cache","User-Agent":"Stripe/1.0 (+https://stripe.com/docs/webhooks)","Accept":"*/*; q=0.5, application/xml","Stripe-Signature":"t=1625667607,v1=2e0492f6578693adce5e581ef3df932c6d733970fbf18ab209a0efb51e081d44,v0=d090bb7ba128744a9b6d2a3898a2f15d08be6d3b2fccf3dfea83b3bb7c3f6207"},"endpoint":{"api_version":"2020-08-27"},"type":"webhook_event"}" prefix=websocket.Client.readPump
time="Wed, 07 Jul 2021 14:20:10 UTC" level=debug msg="Processing webhook event" prefix=proxy.Proxy.processWebhookEvent webhook_converesation_id="1625667607-wc_LvR5FnwsjOWiVG" webhook_id="1625667607-wh_BjYS4iHtNOuaoS"
time="Wed, 07 Jul 2021 14:20:10 UTC" level=debug msg="Received event with non-default API version, ignoring" api_version=2020-08-27 prefix=proxy.Proxy.filterWebhookEvent
time="Wed, 07 Jul 2021 14:20:11 UTC" level=debug msg="Sending ping message" prefix=websocket.Client.writePump
time="Wed, 07 Jul 2021 14:20:11 UTC" level=debug msg="Received pong message" prefix=websocket.Client.readPump
time="Wed, 07 Jul 2021 14:20:12 UTC" level=debug msg="Incoming message" message="{"webhook_id":"1625667607-wh_zy7r5i11RIzuOD","webhook_conversation_id":"1625667607-wc_iLI9nRX1npWE26","event_payload":"{\n  \"id\": \"evt_1JAbbfCJZWINMVesDB7BL9Gd\",\n  \"object\": \"event\",\n  \"api_version\": \"2015-09-03\",\n  \"created\": 1625667607,\n  \"data\": {\n    \"object\": {\n      \"id\": \"sub_Jo9k3qUoCL2CdR\",\n      \"object\": \"subscription\",\n      \"application_fee_percent\": null,\n      \"automatic_tax\": {\n        \"enabled\": false\n      },\n      \"billing\": \"charge_automatically\",\n      \"billing_cycle_anchor\": 1628734656,\n      \"billing_thresholds\": null,\n      \"cancel_at\": null,\n      \"cancel_at_period_end\": false,\n      \"canceled_at\": null,\n      \"collection_method\": \"charge_automatically\",\n      \"created\": 1625651544,\n      \"current_period_end\": 1628734656,\n      \"current_period_start\": 1625651544,\n      \"customer\": \"cus_Jo9KjcY44fGzj1\",\n      \"days_until_due\": null,\n      \"default_payment_method\": null,\n      \"default_source\": null,\n      \"default_tax_rates\": [\n\n      ],\n      \"discount\": null,\n      \"ended_at\": null,\n      \"invoice_customer_balance_settings\": {\n        \"consume_applied_balance_on_void\": true\n      },\n      \"items\": {\n        \"object\": \"list\",\n        \"data\": [\n          {\n            \"id\": \"si_Jo9kt4sglulMkz\",\n            \"object\": \"subscription_item\",\n            \"billing_thresholds\": null,\n            \"created\": 1625651545,\n            \"metadata\": {\n              \"orderCount\": \"1\",\n              \"lastOrderId\": \"1033825\"\n            },\n            \"plan\": {\n              \"id\": \"price_1HpECzCJZWINMVesSprmNjl5\",\n              \"object\": \"plan\",\n              \"active\": true,\n              \"aggregate_usage\": null,\n              \"amount\": 1680,\n              \"amount_decimal\": \"1680\",\n              \"billing_scheme\": \"per_unit\",\n              \"created\": 1612534011,\n              \"currency\": \"gbp\",\n              \"interval\": \"day\",\n              \"interval_count\": 28,\n              \"livemode\": false,\n              \"metadata\": {\n                \"boxes\": \"1\",\n                \"boxSize\": \"28\",\n                \"dailyConsumption\": \"1\",\n                \"baseSku\": \"TURME\",\n                \"retired\": \"0\",\n                \"productVariant\": \"TURME28\"\n              },\n              \"name\": \"2021 Turmeric+ 28\",\n              \"nickname\": \"2021 Turmeric+ 28\",\n              \"product\": \"prod_IQ4LQ4YFeqYB20\",\n              \"statement_descriptor\": \"FUTUREYOU-TURME28\",\n              \"tiers\": null,\n              \"tiers_mode\": null,\n              \"transform_usage\": null,\n              \"trial_period_days\": 24,\n              \"usage_type\": \"licensed\"\n            },\n            \"price\": {\n              \"id\": \"price_1HpECzCJZWINMVesSprmNjl5\",\n              \"object\": \"price\",\n              \"active\": true,\n              \"billing_scheme\": \"per_unit\",\n              \"created\": 1612534011,\n              \"currency\": \"gbp\",\n              \"livemode\": false,\n              \"lookup_key\": null,\n              \"metadata\": {\n                \"boxes\": \"1\",\n                \"boxSize\": \"28\",\n                \"dailyConsumption\": \"1\",\n                \"baseSku\": \"TURME\",\n                \"retired\": \"0\",\n                \"productVariant\": \"TURME28\"\n              },\n              \"nickname\": \"2021 Turmeric+ 28\",\n              \"product\": \"prod_IQ4LQ4YFeqYB20\",\n              \"recurring\": {\n                \"aggregate_usage\": null,\n                \"interval\": \"day\",\n                \"interval_count\": 28,\n                \"trial_period_days\": 24,\n                \"usage_type\": \"licensed\"\n              },\n              \"tiers_mode\": null,\n              \"transform_quantity\": null,\n              \"type\": \"recurring\",\n              \"unit_amount\": 1680,\n              \"unit_amount_decimal\": \"1680\"\n            },\n            \"quantity\": 1,\n            \"subscription\": \"sub_Jo9k3qUoCL2CdR\",\n            \"tax_rates\": [\n\n            ]\n          }\n        ],\n        \"has_more\": false,\n        \"total_count\": 1,\n        \"url\": \"/v1/subscription_items?subscription=sub_Jo9k3qUoCL2CdR\"\n      },\n      \"latest_invoice\": \"in_1JAXQbCJZWINMVes1Sjxxxh6\",\n      \"livemode\": false,\n      \"metadata\": {\n        \"shipping\": \"true\",\n        \"shippingUpgraded\": \"false\",\n        \"source\": \"magento\",\n        \"orderCount\": \"1\",\n        \"lastOrderId\": \"1033825\"\n      },\n      \"next_pending_invoice_item_invoice\": null,\n      \"pause_collection\": null,\n      \"pending_invoice_item_interval\": null,\n      \"pending_setup_intent\": null,\n      \"pending_update\": null,\n      \"plan\": {\n        \"id\": \"price_1HpECzCJZWINMVesSprmNjl5\",\n        \"object\": \"plan\",\n        \"active\": true,\n        \"aggregate_usage\": null,\n        \"amount\": 1680,\n        \"amount_decimal\": \"1680\",\n        \"billing_scheme\": \"per_unit\",\n        \"created\": 1612534011,\n        \"currency\": \"gbp\",\n        \"interval\": \"day\",\n        \"interval_count\": 28,\n        \"livemode\": false,\n        \"metadata\": {\n          \"boxes\": \"1\",\n          \"boxSize\": \"28\",\n          \"dailyConsumption\": \"1\",\n          \"baseSku\": \"TURME\",\n          \"retired\": \"0\",\n          \"productVariant\": \"TURME28\"\n        },\n        \"name\": \"2021 Turmeric+ 28\",\n        \"nickname\": \"2021 Turmeric+ 28\",\n        \"product\": \"prod_IQ4LQ4YFeqYB20\",\n        \"statement_descriptor\": \"FUTUREYOU-TURME28\",\n        \"tiers\": null,\n        \"tiers_mode\": null,\n        \"transform_usage\": null,\n        \"trial_period_days\": 24,\n        \"usage_type\": \"licensed\"\n      },\n      \"quantity\": 1,\n      \"schedule\": null,\n      \"start\": 1625667606,\n      \"start_date\": 1625651544,\n      \"status\": \"trialing\",\n      \"tax_percent\": null,\n      \"transfer_data\": null,\n      \"trial_end\": 1628734656,\n      \"trial_start\": 1625651544\n    },\n    \"previous_attributes\": {\n      \"billing_cycle_anchor\": 1627612809,\n      \"current_period_end\": 1627612809,\n      \"start\": 1625667560,\n      \"trial_end\": 1627612809\n    }\n  },\n  \"livemode\": false,\n  \"pending_webhooks\": 5,\n  \"request\": \"req_oVXbEfceo8VdC1\",\n  \"type\": \"customer.subscription.updated\"\n}","http_headers":{"Content-Type":"application/json; charset=utf-8","Cache-Control":"no-cache","User-Agent":"Stripe/1.0 (+https://stripe.com/docs/webhooks)","Accept":"*/*; q=0.5, application/xml","Stripe-Signature":"t=1625667607,v1=e17b07f1a1235846af14f801623cf4c0b0e0a464184a44298b8112437307a363,v0=e534420e1b43937c0aaeeca81153670e451b7dd26abe014980e578e89a25115d"},"endpoint":{"api_version":null},"type":"webhook_event"}" prefix=websocket.Client.readPump
time="Wed, 07 Jul 2021 14:20:12 UTC" level=debug msg="Processing webhook event" prefix=proxy.Proxy.processWebhookEvent webhook_converesation_id="1625667607-wc_iLI9nRX1npWE26" webhook_id="1625667607-wh_zy7r5i11RIzuOD"
time="Wed, 07 Jul 2021 14:20:12 UTC" level=debug msg="Received malformed event from Stripe, ignoring"
time="Wed, 07 Jul 2021 14:20:13 UTC" level=debug msg="Sending ping message" prefix=websocket.Client.writePump
time="Wed, 07 Jul 2021 14:20:13 UTC" level=debug msg="Received pong message" prefix=websocket.Client.readPump

```
### Environment
docker-compose setup on macOS

### PR DESCRIPTION
**Pull request by suz-stripe**
### Reviewers
r? @vcheung-stripe 
cc @stripe/developer-products

 ### Summary
Fixes #702. 

It turns out that Stripe event payloads belonging to older API versions present the `request` field as a string, instead of a map. This caused our unmarshalling step to soft fail as we ignored any 'invalid' payload shapes. The current tests are passing in the CLI for these changes.

+ [x] ~~I still need to manually verify this with the VS Code extension's RPC pipeline~~

edit: I have verified the RPC pipeline with local CLI build's debug logs, and manually verified in VS Code extension as well. Both new and old event payload shapes are coming through as expected with the correct fields and values 🎉

### CODE REVIEW

File: rpc/commands.pb.go [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
 // 	protoc-gen-go v1.26.0
-// 	protoc        v3.15.8
```
Comment (suz-stripe) [inline rpc/commands.pb.go:4]: I will remove the autogenned files from this PR.

File: pkg/proxy/proxy.go [hunk 2]
```diff
@@ -517,6 +527,43 @@ func Init(cfg *Config) (*Proxy, error) {
...
+	case reflect.String:
+		req = StripeRequest{
+			ID:             data.(string),
+			IdempotencyKey: "",
+		}
+	case reflect.Map:
+		reqDataMap := reqDataValue.Interface().(map[string]interface{})
+
+		var id = ""
+		if reqDataMap["id"] != nil {
```
Comment (vcheung-stripe) [inline pkg/proxy/proxy.go:548]: nit: we've generally been using the two-valued construct for maps in the codebase: https://github.com/stripe/stripe-cli/blob/529cdff8b011fa59fdd872a350b51dfa7e55030e/pkg/fixtures/triggers.go#L112

File: pkg/proxy/proxy.go [hunk 3]
```diff
@@ -517,6 +527,43 @@ func Init(cfg *Config) (*Proxy, error) {
...
 }
 
+// ExtractRequestData takes an interface with request data from a Stripe event payload
+// and properly parses it into a StripeRequest struct before returning it
+func ExtractRequestData(data interface{}) (StripeRequest, error) {
+	var req StripeRequest
+
+	reqDataValue := reflect.ValueOf(data)
+	// here we check the type of the RequestData as it could be either a string or a map
+	// this depends on which API version is in use when listening to events
```
Comment (vcheung-stripe) [inline pkg/proxy/proxy.go:537]: Should we specify which version? It's 2017-05-25 right?

**Review by vcheung-stripe — APPROVED** [review]
LGTM! I'm ok with the naming and I don't have a strong preference for how to deserialize the payload. I just really wish there were polymorphic types in go :(

Just nits so I'm approving 👍

File: pkg/proxy/proxy.go (hunk 3 above)
Comment (suz-stripe) [reply pkg/proxy/proxy.go:537]: 👍  great idea to document, and it's all versions including and before that I believe

File: pkg/proxy/proxy.go (hunk 2 above)
Comment (suz-stripe) [reply pkg/proxy/proxy.go:548]: thank you for this callout! Will update

**Pull request merged**